drivers/fuel_gauge/bq27z746/bq27z746.c

```diff
@@ -0,0 +1,324 @@
+{
+	if (len > BQ27Z746_MAC_DATA_LEN) {
+		return -EINVAL;
+	}
+
+	uint8_t buf[BQ27Z746_MAC_COMPLETE_LEN];
+	const struct bq27z746_config *cfg = dev->config;
+
+	/* Instead of MAC, ALTMAC is used as reccommended in the datasheet */
+	int ret = bq27z746_write16(dev, BQ27Z746_ALTMANUFACTURERACCESS, cmd);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	/*
+	 * The data read from BQ27Z746_ALTMANUFACTURERACCESS is:
+	 * 0..1: The command (for verification)
+	 * 2..33: The data
+	 * 34: Checksum calculated as (uint8_t)(0xFF - (sum of all command and data bytes))
+	 * 35: Length including command, checksum and length (e.g. data length + 4)
+	 */
+
+	ret = i2c_burst_read_dt(&cfg->i2c, BQ27Z746_ALTMANUFACTURERACCESS, buf,
+				BQ27Z746_MAC_COMPLETE_LEN);
+	if (ret != 0) {
+		return ret;
+	}
+
+	/* The first two bytes read is the command and is used for verification */
+	const uint16_t read_cmd = sys_get_le16(buf);
+
+	if (read_cmd != cmd) {
+		LOG_ERR("Read command 0x%x != written command 0x%x", read_cmd, cmd);
+		return -EIO;
+	}
+
+	const uint8_t checksum_actual = buf[34];
+	uint8_t sum = 0; /* Intentionally 8 bit wide and overflowing */
+
+	for (int i = 0; i < BQ27Z746_MAC_COMPLETE_LEN - 2; i++) {
+		sum += buf[i];
+	}
+
+	const uint8_t checksum_expected = 0xFF - sum;
+
+	if (checksum_expected != checksum_actual) {
+		LOG_ERR("Checksum mismatch");
+		return -EIO;
+	}
+
+	/* First byte of the user buffer is the length */
+	data[0] = buf[35] - BQ27Z746_MAC_OVERHEAD_LEN;
+
+	/* Copy only the data to the user buffer (= skipping the first two command bytes) */
+	memcpy(&data[1], &buf[2], len);
+
+	return ret;
+}
+
+static int bq27z746_get_prop(const struct device *dev, struct fuel_gauge_get_property *prop)
+{
+	int rc = 0;
+	uint16_t val = 0;
+
+	/*
+	 * Possibly negative values must be cast from uint16 to int16 first to
+	 * then correctly end up in the wider datatypes of `prop`.
+	 */
+
+	switch (prop->property_type) {
+	case FUEL_GAUGE_AVG_CURRENT:
+		rc = bq27z746_read16(dev, BQ27Z746_AVERAGECURRENT, &val);
+		prop->value.avg_current = (int16_t)val * 1000;
+		break;
+	case FUEL_GAUGE_CYCLE_COUNT:
+		rc = bq27z746_read16(dev, BQ27Z746_CYCLECOUNT, &val);
+		prop->value.cycle_count = val * 100;
+		break;
+	case FUEL_GAUGE_CURRENT:
+		rc = bq27z746_read16(dev, BQ27Z746_CURRENT, &val);
+		prop->value.current = (int16_t)val * 1000;
+		break;
+	case FUEL_GAUGE_FULL_CHARGE_CAPACITY:
+		rc = bq27z746_read16(dev, BQ27Z746_FULLCHARGECAPACITY, &val);
+		prop->value.full_charge_capacity = val * 1000;
+		break;
+	case FUEL_GAUGE_REMAINING_CAPACITY:
+		rc = bq27z746_read16(dev, BQ27Z746_REMAININGCAPACITY, &val);
+		prop->value.remaining_capacity = val * 1000;
+		break;
+	case FUEL_GAUGE_RUNTIME_TO_EMPTY:
+		rc = bq27z746_read16(dev, BQ27Z746_AVERAGETIMETOEMPTY, &val);
+		prop->value.runtime_to_empty = val;
+		break;
+	case FUEL_GAUGE_RUNTIME_TO_FULL:
+		rc = bq27z746_read16(dev, BQ27Z746_AVERAGETIMETOFULL, &val);
+		prop->value.runtime_to_empty = val;
```

Comment:
this should be RTTF

Reply:
Good catch, will fix it to avoid confusion.

Reply:
+1 Good catch!

Reply:
Fixed @perowle